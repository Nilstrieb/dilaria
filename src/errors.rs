//!
//! This modules handles error reporting in the interpreter

use std::fmt::Debug;

pub use span::Span;

mod span {

    #[derive(Debug, Default, Copy, Clone, PartialOrd, PartialEq, Ord, Eq, Hash)]
    pub struct Span {
        pub start: usize,
        pub end: usize,
    }

    impl Span {
        pub fn start_len(start: usize, len: usize) -> Self {
            Self {
                start,
                end: start + len,
            }
        }

        pub fn start_end(start: usize, end: usize) -> Self {
            Self { start, end }
        }

        pub fn single(start: usize) -> Self {
            Self {
                start,
                end: start + 1,
            }
        }

        pub fn dummy() -> Self {
            Self { start: 0, end: 0 }
        }

        /// Extends the span by the second one
        /// The other one has to be after the current one
        pub fn extend(&self, other: Span) -> Span {
            debug_assert!(self.start <= other.start);
            debug_assert!(self.end <= other.end);
            Span {
                start: self.start,
                end: other.end,
            }
        }

        /// Extends the span by the second one, if it exists
        /// The other one has to be after the current one, if it exists
        pub fn option_extend(&self, other: Option<Span>) -> Span {
            match other {
                None => *self,
                Some(span) => self.extend(span),
            }
        }

        pub fn len(&self) -> usize {
            self.end - self.start
        }
    }
}

#[derive(Debug, Clone)]
pub struct CompilerError {
    pub span: Span,
    pub message: String,
    pub note: Option<String>,
}

impl CompilerError {
    pub fn new(span: Span, message: String) -> Self {
        Self {
            span,
            message,
            note: None,
        }
    }

    pub fn with_note(span: Span, message: String, note: String) -> Self {
        Self {
            span,
            message,
            note: Some(note),
        }
    }
}

pub fn display_error(source: &str, error: CompilerError) {
    let span = error.span;

    let mut chars = 0;
    let lines = source.split_inclusive('\n').enumerate();
    for (idx, line) in lines {
        if chars + line.len() > span.start {
            let offset_on_line = span.start - chars;

            println!("{}error: {}{}", RED, error.message, RESET);
            println!("      {}|{}", CYAN, RESET);
            println!("{}{:>5} |{} {}", CYAN, idx + 1, RESET, line);
            print!("      {}|{} ", CYAN, RESET);
            println!(
                "{}{}{}{}",
                " ".repeat(offset_on_line),
                RED,
                "^".repeat(span.len()),
                RESET,
            );
            if let Some(note) = error.note {
                println!("      {}|{}", CYAN, RESET);
                println!(
                    "      {}|{}   {}note: {}{}",
                    CYAN, RESET, GREEN, note, RESET
                );
            }
            break;
        }
        chars += line.len();
    }
}

macro_rules! color {
    ($name:ident: $value:literal) => {
        const $name: &str = concat!("\x1B[", $value);
    };
}

color!(RED: "0;31m");
color!(RESET: "0m");
color!(CYAN: "0;36m");
color!(GREEN: "0;32m");
