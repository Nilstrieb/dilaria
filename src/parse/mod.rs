#![allow(dead_code)]

#[cfg(test)]
mod test;

use crate::ast::*;
use crate::errors::{CompilerError, Span};
use crate::lex::{Token, TokenType};
use std::iter::Peekable;

pub fn parse(tokens: Vec<Token>) -> Result<Program, ParseErr> {
    let mut parser = Parser {
        tokens: tokens.into_iter().peekable(),
        inside_fn_depth: 0,
        inside_loop_depth: 0,
    };
    let program = parser.program()?;
    Ok(program)
}

#[derive(Debug)]
struct Parser<'code> {
    tokens: Peekable<std::vec::IntoIter<Token<'code>>>,
    inside_fn_depth: usize,
    inside_loop_depth: usize,
}

type ParseResult<'code, T> = Result<T, ParseErr<'code>>;

macro_rules! parse_bin_op {
    ($self: ident, $lhs: ident, $kind: expr, $function: ident) => {{
        let _ = $self.next();
        let rhs = $self.$function()?;
        Ok(Expr::BinaryOp(Box::new(BinaryOp {
            span: $lhs.span().extend(rhs.span()),
            lhs: $lhs,
            rhs,
            kind: $kind,
        })))
    }};
}

impl<'code> Parser<'code> {
    fn program(&mut self) -> ParseResult<'code, Program> {
        Ok(Program(self.statement_list()?))
    }

    fn statement_list(&mut self) -> ParseResult<'code, Vec<Stmt>> {
        let mut stmts = Vec::new();
        loop {
            if let Some(TokenType::BraceC) | None = self.peek_kind() {
                let _ = self.next();
                return Ok(stmts);
            }
            let stmt = self.statement()?;
            stmts.push(stmt);
        }
    }

    fn block(&mut self) -> ParseResult<'code, Block> {
        let start_span = self.expect(TokenType::BraceO)?.span;
        let stmts = self.statement_list()?;
        let end_span = self.expect(TokenType::BraceC)?.span;
        Ok(Block {
            stmts,
            span: start_span.extend(end_span),
        })
    }

    fn statement(&mut self) -> ParseResult<'code, Stmt> {
        match self.peek_kind().ok_or(ParseErr::EOF("statement"))? {
            &TokenType::Let => self.declaration(),
            &TokenType::Fn => self.fn_decl(),
            &TokenType::If => Ok(Stmt::If(self.if_stmt()?)),
            &TokenType::Loop => self.loop_stmt(),
            &TokenType::While => self.while_stmt(),
            &TokenType::Break => self.break_stmt(),
            &TokenType::Return => self.return_stmt(),
            &TokenType::BraceO => Ok(Stmt::Block(self.block()?)),
            _ => {
                let expr = self.expression()?;
                self.expect(TokenType::Semi)?;
                Ok(Stmt::Expr(expr))
            }
        }
    }

    fn declaration(&mut self) -> ParseResult<'code, Stmt> {
        todo!()
    }

    fn assignment(&mut self) -> ParseResult<'code, Stmt> {
        todo!()
    }

    fn fn_decl(&mut self) -> ParseResult<'code, Stmt> {
        todo!()
    }

    fn if_stmt(&mut self) -> ParseResult<'code, IfStmt> {
        let keyword_span = self.expect(TokenType::If)?.span;
        let cond = self.expression()?;
        let body = self.block()?;

        let else_part = if let Some(TokenType::Else) = self.peek_kind() {
            Some(self.else_part()?)
        } else {
            None
        };

        Ok(IfStmt {
            span: keyword_span
                .extend(body.span)
                .option_extend(else_part.as_ref().map(|part| part.span())),
            cond,
            body,
            else_part: else_part.map(Box::new),
        })
    }

    fn else_part(&mut self) -> ParseResult<'code, ElsePart> {
        let keyword_span = self.expect(TokenType::Else)?.span;

        if let Some(TokenType::If) = self.peek_kind() {
            let else_if_stmt = self.if_stmt()?;
            let else_span = keyword_span.extend(else_if_stmt.span);
            Ok(ElsePart::ElseIf(else_if_stmt, else_span))
        } else {
            let block = self.block()?;
            let else_span = keyword_span.extend(block.span);
            Ok(ElsePart::Else(block, else_span))
        }
    }

    fn loop_stmt(&mut self) -> ParseResult<'code, Stmt> {
        let keyword_span = self.expect(TokenType::Loop)?.span;

        self.inside_loop_depth += 1;
        let block = self.block()?;
        self.inside_loop_depth -= 1;

        let loop_span = keyword_span.extend(block.span);
        Ok(Stmt::Loop(block, keyword_span.extend(loop_span)))
    }

    fn while_stmt(&mut self) -> ParseResult<'code, Stmt> {
        let keyword_span = self.expect(TokenType::While)?.span;
        let cond = self.expression()?;
        let body = self.block()?;
        Ok(Stmt::While(WhileStmt {
            span: keyword_span.extend(body.span),
            cond,
            body,
        }))
    }

    fn break_stmt(&mut self) -> ParseResult<'code, Stmt> {
        let keyword_span = self.expect(TokenType::Break)?.span;
        let semi_span = self.expect(TokenType::Semi)?.span;

        if self.inside_loop_depth == 0 {
            Err(ParseErr::BreakOutsideLoop(keyword_span.extend(semi_span)))
        } else {
            Ok(Stmt::Break(keyword_span.extend(semi_span)))
        }
    }

    fn return_stmt(&mut self) -> ParseResult<'code, Stmt> {
        let keyword_span = self.expect(TokenType::Return)?.span;

        let expr = if let Some(TokenType::Semi) = self.peek_kind() {
            None
        } else {
            Some(self.expression()?)
        };

        let semi_span = self.expect(TokenType::Semi)?.span;

        if self.inside_fn_depth == 0 {
            Err(ParseErr::ReturnOutsideFunction(
                keyword_span.extend(semi_span),
            ))
        } else {
            Ok(Stmt::Return(expr, keyword_span.extend(semi_span)))
        }
    }

    fn expression(&mut self) -> ParseResult<'code, Expr> {
        self.logical_or()
    }

    fn logical_or(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.logical_and()?;
        match self.peek_kind() {
            Some(TokenType::Or) => parse_bin_op!(self, lhs, BinaryOpKind::Or, logical_and),
            _ => Ok(lhs),
        }
    }

    fn logical_and(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.equality()?;
        match self.peek_kind() {
            Some(TokenType::And) => parse_bin_op!(self, lhs, BinaryOpKind::And, equality),
            _ => Ok(lhs),
        }
    }

    fn equality(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.comparison()?;
        match self.peek_kind() {
            Some(TokenType::BangEqual) => {
                parse_bin_op!(self, lhs, BinaryOpKind::NotEqual, comparison)
            }
            Some(TokenType::EqualEqual) => {
                parse_bin_op!(self, lhs, BinaryOpKind::Equal, comparison)
            }
            _ => Ok(lhs),
        }
    }

    fn comparison(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.term()?;
        match self.peek_kind() {
            Some(TokenType::Greater) => parse_bin_op!(self, lhs, BinaryOpKind::Greater, term),
            Some(TokenType::GreaterEqual) => {
                parse_bin_op!(self, lhs, BinaryOpKind::GreaterEqual, term)
            }
            Some(TokenType::Less) => parse_bin_op!(self, lhs, BinaryOpKind::Less, term),
            Some(TokenType::LessEqual) => {
                parse_bin_op!(self, lhs, BinaryOpKind::LessEqual, term)
            }
            _ => Ok(lhs),
        }
    }

    fn term(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.factor()?;
        match self.peek_kind() {
            Some(TokenType::Plus) => parse_bin_op!(self, lhs, BinaryOpKind::Add, factor),
            Some(TokenType::Minus) => parse_bin_op!(self, lhs, BinaryOpKind::Sub, factor),
            _ => Ok(lhs),
        }
    }

    fn factor(&mut self) -> ParseResult<'code, Expr> {
        let lhs = self.unary()?;
        match self.peek_kind() {
            Some(TokenType::Asterisk) => parse_bin_op!(self, lhs, BinaryOpKind::Mul, unary),
            Some(TokenType::Slash) => parse_bin_op!(self, lhs, BinaryOpKind::Div, unary),
            Some(TokenType::Percent) => parse_bin_op!(self, lhs, BinaryOpKind::Mod, unary),
            _ => Ok(lhs),
        }
    }

    fn unary(&mut self) -> ParseResult<'code, Expr> {
        match self.peek_kind() {
            Some(TokenType::Not) => {
                let unary_op_span = self.next().unwrap().span;
                let expr = self.expression()?;
                Ok(Expr::UnaryOp(Box::new(UnaryOp {
                    span: unary_op_span.extend(expr.span()),
                    expr,
                    kind: UnaryOpKind::Not,
                })))
            }
            Some(TokenType::Minus) => {
                let unary_op_span = self.next().unwrap().span;
                let expr = self.expression()?;
                Ok(Expr::UnaryOp(Box::new(UnaryOp {
                    span: unary_op_span.extend(expr.span()),
                    expr,
                    kind: UnaryOpKind::Neg,
                })))
            }
            _ => self.primary(),
        }
    }

    fn primary(&mut self) -> ParseResult<'code, Expr> {
        let next = self.next().ok_or(ParseErr::EOF("primary"))?;
        match next.kind {
            TokenType::String(literal) => Ok(Expr::Literal(Literal::String(literal, next.span))),
            TokenType::Number(literal) => Ok(Expr::Literal(Literal::Number(literal, next.span))),
            TokenType::False => Ok(Expr::Literal(Literal::Boolean(false, next.span))),
            TokenType::True => Ok(Expr::Literal(Literal::Boolean(true, next.span))),
            TokenType::Null => Ok(Expr::Literal(Literal::Null(next.span))),
            TokenType::BraceO => self.object_literal(next.span),
            TokenType::BracketO => self.array_literal(next.span),
            TokenType::ParenO => {
                let expr = self.expression()?;
                let _ = self.expect(TokenType::ParenC)?;
                Ok(expr)
            }
            TokenType::Ident(name) => {
                let name_owned = name.to_owned();
                Ok(Expr::Ident(name_owned, next.span))
            }
            _ => Err(ParseErr::InvalidTokenPrimary(next)),
        }
    }

    fn object_literal(&mut self, open_span: Span) -> ParseResult<'code, Expr> {
        let close_span = self.expect(TokenType::BraceC)?.span;
        Ok(Expr::Literal(Literal::Object(open_span.extend(close_span))))
    }

    fn array_literal(&mut self, open_span: Span) -> ParseResult<'code, Expr> {
        let mut elements = Vec::new();
        while self
            .peek()
            .ok_or(ParseErr::EOFExpecting(TokenType::BracketC))?
            .kind
            != TokenType::BracketC
        {
            let expr = self.expression()?;
            elements.push(expr);
            self.expect(TokenType::Comma)?;
        }
        let closing_bracket = self.expect(TokenType::BracketC)?;
        Ok(Expr::Literal(Literal::Array(
            elements,
            open_span.extend(closing_bracket.span),
        )))
    }

    // token helpers

    #[must_use]
    fn next(&mut self) -> Option<Token<'code>> {
        self.tokens.next()
    }

    #[must_use]
    fn peek(&mut self) -> Option<&Token<'code>> {
        self.tokens.peek()
    }

    #[must_use]
    fn peek_kind(&mut self) -> Option<&TokenType<'code>> {
        self.peek().map(|token| &token.kind)
    }

    fn expect(&mut self, kind: TokenType<'code>) -> ParseResult<'code, Token> {
        if let Some(token) = self.next() {
            if token.kind == kind {
                Ok(token)
            } else {
                Err(ParseErr::MismatchedKind {
                    expected: kind,
                    actual: token,
                })
            }
        } else {
            Err(ParseErr::EOFExpecting(kind))
        }
    }
}

#[derive(Debug)]
pub enum ParseErr<'code> {
    BreakOutsideLoop(Span),
    ReturnOutsideFunction(Span),
    MismatchedKind {
        expected: TokenType<'code>,
        actual: Token<'code>,
    },
    InvalidTokenPrimary(Token<'code>),
    EOFExpecting(TokenType<'code>),
    EOF(&'static str),
}

impl CompilerError for ParseErr<'_> {
    fn span(&self) -> Span {
        match self {
            ParseErr::MismatchedKind {
                actual: Token { span, .. },
                ..
            } => *span,
            ParseErr::InvalidTokenPrimary(Token { span, .. }) => *span,
            ParseErr::EOFExpecting(_) => Span::dummy(),
            ParseErr::EOF(_) => Span::dummy(),
            ParseErr::BreakOutsideLoop(span) => *span,
            ParseErr::ReturnOutsideFunction(span) => *span,
        }
    }

    fn message(&self) -> String {
        match self {
            ParseErr::MismatchedKind { expected, actual } => {
                format!("expected: {:?}, received: {:?}", expected, actual.kind)
            }
            ParseErr::InvalidTokenPrimary(token) => {
                format!("invalid token in expression: {:?}", token.kind)
            }
            ParseErr::EOFExpecting(token) => {
                format!("reached EOF searching for: {:?}", token)
            }
            ParseErr::EOF(message) => {
                format!("reached EOF while parsing: {}", message)
            }
            ParseErr::BreakOutsideLoop(_) => "break used outside of loop".to_string(),
            ParseErr::ReturnOutsideFunction(_) => "return used outside of function".to_string(),
        }
    }

    fn note(&self) -> Option<String> {
        None
    }
}
