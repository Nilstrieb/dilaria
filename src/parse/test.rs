use crate::errors::Span;
use crate::parse::Parser;
use prelude::*;

mod prelude {
    pub(super) use super::{parser, test_literal_bin_op, test_number_literal, token};
    pub(super) use crate::ast::{Expr, Stmt};
    pub(super) use crate::lex::{
        Token,
        TokenType::{self, *},
    };
}

fn token(kind: TokenType) -> Token {
    Token {
        span: Span::dummy(),
        kind,
    }
}

fn parser(tokens: Vec<Token>) -> Parser {
    Parser {
        tokens: tokens.into_iter().peekable(),
        depth: 0,
        inside_fn_depth: 0,
        inside_loop_depth: 0,
    }
}

fn test_literal_bin_op<F: FnOnce(Vec<Token<'_>>) -> Expr>(token_type: TokenType, parser: F) {
    let tokens = [Number(10.0), token_type, Number(4.0)].map(token).into();
    let ast = parser(tokens);
    insta::assert_debug_snapshot!(ast);
}

fn test_number_literal<F: FnOnce(Vec<Token<'_>>) -> Expr>(parser: F) {
    let tokens = [Number(10.0)].map(token).into();
    let ast = parser(tokens);
    insta::assert_debug_snapshot!(ast);
}

mod assignment {
    use super::prelude::*;

    fn parse_assignment(tokens: Vec<Token>) -> Stmt {
        let mut parser = parser(tokens);
        parser.assignment().unwrap()
    }

    #[test]
    fn simple() {
        let tokens = [Ident("hugo"), Equal, Number(10.0), Semi].map(token).into();
        let ast = parse_assignment(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn call_expr() {
        let tokens = [
            Ident("hugo"),
            Dot,
            Ident("age"),
            Equal,
            Number(2021.0),
            Minus,
            Number(1986.0),
            Semi,
        ]
        .map(token)
        .into();
        let ast = parse_assignment(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod r#fn {
    use super::prelude::*;

    fn parse_fn(tokens: Vec<Token>) -> Stmt {
        let mut parser = parser(tokens);
        parser.fn_decl().unwrap()
    }

    #[test]
    fn empty() {
        let tokens = [Fn, Ident("empty"), ParenO, ParenC, BraceO, BraceC]
            .map(token)
            .into();
        let ast = parse_fn(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn params_body() {
        let tokens = [
            Fn,
            Ident("empty"),
            ParenO,
            Ident("a"),
            Comma,
            Ident("b"),
            ParenC,
            BraceO,
            Number(10.0),
            Plus,
            Number(20.0),
            Semi,
            BraceC,
        ]
        .map(token)
        .into();
        let ast = parse_fn(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod r#if {
    use super::prelude::*;
    use crate::ast::IfStmt;

    fn parse_if(tokens: Vec<Token>) -> IfStmt {
        let mut parser = parser(tokens);
        parser.if_stmt().unwrap()
    }

    #[test]
    fn empty() {
        let tokens = [If, True, BraceO, BraceC].map(token).into();
        let ast = parse_if(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn if_else() {
        let tokens = [If, True, BraceO, BraceC, Else, BraceO, BraceC]
            .map(token)
            .into();
        let ast = parse_if(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn if_else_if() {
        let tokens = [If, True, BraceO, BraceC, Else, If, True, BraceO, BraceC]
            .map(token)
            .into();
        let ast = parse_if(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn if_else_if_else() {
        let tokens = [
            If, True, BraceO, BraceC, Else, If, True, BraceO, BraceC, Else, BraceO, BraceC,
        ]
        .map(token)
        .into();
        let ast = parse_if(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod print {
    use super::prelude::*;

    fn parse_print(tokens: Vec<Token>) -> Stmt {
        let mut parser = parser(tokens);
        parser.print_stmt().unwrap()
    }

    #[test]
    fn print_true() {
        let tokens = [Print, True, Semi].map(token).into();
        let ast = parse_print(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod r#while {
    use super::prelude::*;

    fn parse_while(tokens: Vec<Token>) -> Stmt {
        let mut parser = parser(tokens);
        parser.while_stmt().unwrap()
    }

    #[test]
    fn empty() {
        let tokens = [While, True, BraceO, BraceC].map(token).into();
        let ast = parse_while(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn or_condition_break() {
        let tokens = [While, False, Or, True, BraceO, Break, Semi, BraceC]
            .map(token)
            .into();
        let ast = parse_while(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod r#loop {
    use super::prelude::*;

    fn parse_loop(tokens: Vec<Token>) -> Stmt {
        let mut parser = parser(tokens);
        parser.loop_stmt().unwrap()
    }

    #[test]
    fn empty() {
        let tokens = [Loop, BraceO, BraceC].map(token).into();
        let ast = parse_loop(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn with_break() {
        let tokens = [Loop, BraceO, Break, Semi, BraceC].map(token).into();
        let ast = parse_loop(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn break_after_inner() {
        let tokens = [Loop, BraceO, Loop, BraceO, BraceC, Break, Semi, BraceC]
            .map(token)
            .into();
        let ast = parse_loop(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod block {
    use super::prelude::*;
    use crate::ast::Block;

    fn parse_block(tokens: Vec<Token>) -> Block {
        let mut parser = parser(tokens);
        parser.block().unwrap()
    }

    #[test]
    fn empty() {
        let tokens = [BraceO, BraceC].map(token).into();
        let ast = parse_block(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn two_expressions() {
        let tokens = [BraceO, Number(10.0), Semi, Number(20.0), Semi, BraceC]
            .map(token)
            .into();
        let ast = parse_block(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn nested() {
        let tokens = [BraceO, BraceO, BraceC, BraceC].map(token).into();
        let ast = parse_block(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod expr {
    use super::prelude::*;

    fn parse_expr(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.expression().unwrap()
    }

    #[test]
    fn stack_overflow() {
        let tokens = std::iter::repeat(BracketO)
            .map(token)
            .take(100_000)
            .collect();
        let expr = parser(tokens).expression();
        assert!(expr.is_err());
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_expr);
    }

    #[test]
    fn add_multiply() {
        let tokens = [Number(10.0), Plus, Number(20.0), Asterisk, Number(100.0)]
            .map(token)
            .into();
        let ast = parse_expr(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn equal_unary() {
        let tokens = [Number(10.0), EqualEqual, Minus, Number(10.0)]
            .map(token)
            .into();
        let ast = parse_expr(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn parentheses_mul_add() {
        let tokens = [
            Number(10.0),
            Asterisk,
            ParenO,
            Number(20.0),
            Plus,
            Number(30.0),
            ParenC,
        ]
        .map(token)
        .into();
        let ast = parse_expr(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod logical_or {
    use super::prelude::*;

    fn parse_logical_or(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.logical_or().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_logical_or);
    }

    #[test]
    fn and() {
        test_literal_bin_op(Or, parse_logical_or);
    }
}

mod logical_and {
    use super::prelude::*;

    fn parse_logical_and(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.logical_and().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_logical_and);
    }

    #[test]
    fn and() {
        test_literal_bin_op(And, parse_logical_and);
    }
}

mod equality {
    use super::prelude::*;

    fn parse_equality(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.equality().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_equality);
    }

    #[test]
    fn not_equal() {
        test_literal_bin_op(BangEqual, parse_equality);
    }

    #[test]
    fn equal() {
        test_literal_bin_op(EqualEqual, parse_equality);
    }
}

mod comparison {
    use super::prelude::*;

    fn parse_comparison(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.comparison().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_comparison);
    }

    #[test]
    fn greater() {
        test_literal_bin_op(Greater, parse_comparison);
    }

    #[test]
    fn greater_equal() {
        test_literal_bin_op(GreaterEqual, parse_comparison);
    }

    #[test]
    fn less() {
        test_literal_bin_op(Less, parse_comparison);
    }

    #[test]
    fn less_equal() {
        test_literal_bin_op(LessEqual, parse_comparison);
    }
}

mod term {
    use super::prelude::*;

    fn parse_term(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.term().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_term);
    }

    #[test]
    fn add() {
        test_literal_bin_op(Plus, parse_term);
    }

    #[test]
    fn sub() {
        test_literal_bin_op(Minus, parse_term);
    }
}

mod factor {
    use super::prelude::*;

    fn parse_factor(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.factor().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_factor);
    }

    #[test]
    fn multiply() {
        test_literal_bin_op(Asterisk, parse_factor);
    }

    #[test]
    fn divide() {
        test_literal_bin_op(Slash, parse_factor);
    }

    #[test]
    fn modulo() {
        test_literal_bin_op(Percent, parse_factor);
    }
}

mod unary {
    use super::prelude::*;

    fn parse_unary(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.unary().unwrap()
    }

    #[test]
    fn number_literal() {
        test_number_literal(parse_unary);
    }

    #[test]
    fn not() {
        let tokens = [Not, True].map(token).into();
        let ast = parse_unary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn neg() {
        let tokens = [Minus, Number(10.0)].map(token).into();
        let ast = parse_unary(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod call {
    use super::prelude::*;

    fn parse_call(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.call().unwrap()
    }

    #[test]
    fn field_simple() {
        let tokens = [Ident("hugo"), Dot, Ident("name")].map(token).into();
        let ast = parse_call(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn simple() {
        let tokens = [Ident("print"), ParenO, ParenC].map(token).into();
        let ast = parse_call(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn fn_args() {
        let tokens = [
            Ident("print"),
            ParenO,
            Number(10.0),
            Comma,
            Number(5.0),
            Comma,
            ParenC,
        ]
        .map(token)
        .into();
        let ast = parse_call(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn nested() {
        let tokens = [
            Ident("hugo"),
            Dot,
            Ident("name"),
            Dot,
            Ident("print"),
            ParenO,
            ParenC,
        ]
        .map(token)
        .into();
        let ast = parse_call(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn with_exprs() {
        // print((10 + 5).abs())
        let tokens = [
            Ident("print"),
            ParenO,
            ParenO,
            Number(10.0),
            Plus,
            Number(5.0),
            ParenC,
            Dot,
            Ident("abs"),
            ParenO,
            ParenC,
            ParenC,
        ]
        .map(token)
        .into();
        let ast = parse_call(tokens);
        insta::assert_debug_snapshot!(ast);
    }
}

mod primary {
    use super::prelude::*;

    fn parse_primary(tokens: Vec<Token>) -> Expr {
        let mut parser = parser(tokens);
        parser.primary().unwrap()
    }

    #[test]
    fn ident_test() {
        let tokens = [Ident("tokens")].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn string() {
        let tokens = [Number(10.0)].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn number() {
        let tokens = [String("uwu".to_string())].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn empty_object() {
        let tokens = [BraceO, BraceC].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn empty_array() {
        let tokens = [BracketO, BracketC].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn r#false() {
        let tokens = [False].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn r#true() {
        let tokens = [True].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn null() {
        let tokens = [Null].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn empty_array_literal() {
        let tokens = [BracketO, BracketC].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn single_array_literal() {
        let tokens = [BracketO, Number(10.0), BracketC].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn single_array_literal_trailing_comma() {
        let tokens = [BracketO, Number(10.0), Comma, BracketC].map(token).into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn two_array_literal() {
        let tokens = [BracketO, Number(10.0), Comma, Number(10.0), BracketC]
            .map(token)
            .into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn two_array_literal_trailing_comma() {
        let tokens = [BracketO, Number(10.0), Comma, Number(10.0), Comma, BracketC]
            .map(token)
            .into();
        let ast = parse_primary(tokens);
        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn two_array_literal_no_comma() {
        let tokens = [BracketO, Number(10.0), Number(10.0), BracketC]
            .map(token)
            .into();
        let mut parser = parser(tokens);
        let expr = parser.primary();
        assert!(expr.is_err());
    }
}
