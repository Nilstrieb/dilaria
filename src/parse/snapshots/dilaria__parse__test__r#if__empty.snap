---
source: src/parse/test.rs
expression: ast
---
IfStmt {
    span: Span(
        0..0,
    ),
    cond: Literal(
        Boolean(
            true,
            Span(
                0..0,
            ),
        ),
    ),
    body: Block {
        stmts: [],
        span: Span(
            0..0,
        ),
    },
    else_part: None,
}
